lib/promscrape/scrapework.go

```diff
@@ -527,51 +526,58 @@ func (sw *scrapeWork) processDataOneShot(scrapeTimestamp, realTimestamp int64, b
 }
 
 func (sw *scrapeWork) processDataInStreamMode(scrapeTimestamp, realTimestamp int64, body *bytesutil.ByteBuffer, scrapeDurationSeconds float64) error {
-	samplesScraped := 0
-	samplesPostRelabeling := 0
-	wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
+	var samplesScraped atomic.Int64
+	var samplesPostRelabeling atomic.Int64
+	var samplesDroppedTotal atomic.Int64
 
 	lastScrape := sw.loadLastScrape()
 	bodyString := bytesutil.ToUnsafeString(body.B)
 	areIdenticalSeries := sw.areIdenticalSeries(lastScrape, bodyString)
-	samplesDropped := 0
 
 	r := body.NewReader()
-	var mu sync.Mutex
 	err := stream.Parse(r, scrapeTimestamp, "", false, func(rows []parser.Row) error {
-		mu.Lock()
-		defer mu.Unlock()
+		var err error
+
+		wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
+		defer func() {
+			wc.resetNoRows()
+			writeRequestCtxPool.Put(wc)
+		}()
+		samplesScraped.Add(int64(len(rows)))
 
-		samplesScraped += len(rows)
 		for i := range rows {
 			sw.addRowToTimeseries(wc, &rows[i], scrapeTimestamp, true)
 		}
-		samplesPostRelabeling += len(wc.writeRequest.Timeseries)
-		if sw.Config.SampleLimit > 0 && samplesPostRelabeling > sw.Config.SampleLimit {
+		newSamplesPostRelabeling := samplesPostRelabeling.Add(int64(len(wc.writeRequest.Timeseries)))
+		if sw.Config.SampleLimit > 0 && int(newSamplesPostRelabeling) > sw.Config.SampleLimit {
 			wc.resetNoRows()
 			scrapesSkippedBySampleLimit.Inc()
-			return fmt.Errorf("the response from %q exceeds sample_limit=%d; "+
+			err = fmt.Errorf("the response from %q exceeds sample_limit=%d; "+
```

Comment:
Auto metrics related to the series limiter are not the same. One shot adds them, where stream is not. 

This change aligns stream behavior with a one-shot one.